ops/model.py

```diff
@@ -3613,69 +3602,16 @@ def __init__(
         self._is_leader: bool | None = None
         self._leader_check_time = None
         self._hook_is_running = ''
-        self._is_recursive = contextvars.ContextVar('_is_recursive', default=False)
 
     @contextlib.contextmanager
-    def _prevent_recursion(self):
-        token = self._is_recursive.set(True)
-        try:
-            yield
-        finally:
-            self._is_recursive.reset(token)
-
-    def _run(
-        self,
-        *args: str,
-        return_output: bool = False,
-        use_json: bool = False,
-        input_stream: str | None = None,
-    ) -> str | Any | None:
-        if self._is_recursive.get():
-            # Either `juju-log` hook tool failed or there's a bug in ops.
-            return
-        # Logs are collected via log integration, omit the subprocess calls that push
-        # the same content to juju from telemetry.
-        mgr = (
-            self._prevent_recursion()
-            if args[0] == 'juju-log'
-            else tracer.start_as_current_span(args[0])
-        )
-        with mgr as span:
-            kwargs = {
-                'stdout': subprocess.PIPE,
-                'stderr': subprocess.PIPE,
-                'check': True,
-                'encoding': 'utf-8',
-            }
-            if input_stream:
-                kwargs.update({'input': input_stream})
-            which_cmd = shutil.which(args[0])
-            if which_cmd is None:
-                raise RuntimeError(f'command not found: {args[0]}')
-            args = (which_cmd,) + args[1:]
-            if use_json:
-                args += ('--format=json',)
+    def _tracing_span(self, cmd: str, *args: Any):
```

Comment:
So the need for the recursion prevention goes away completely, because hookcmds never logs? That's a nice side simplification!

Reply:
I wonder about that.
I'll have to refresh my reading of my own changes from just 6 months ago, but here's the other concern (separate from my comment about spans above):

- something is logged
   - we call `juju-log`
      - which fails, and that's logged
         - we call `juju-log`
            - ...

Reply:
If the failure is an exception, then this is the same as the situation above, it doesn't really matter whether `juju-log` or something else is raising the exception.

Other than the exception case, there's no logging in `hookcmds`, so it can't trigger any `_ModelBackend.juju_log` usage.